### PR TITLE
[RDY] Extend 'generic_factory' + aliases + use the factory to manage terrain objects

### PR DESCRIPTION
**Pull request by codemime**
Continues #15977.
- The `generic_factory` class became responsible for routine id operations (converting, validating, caching (b393d7c), checking  (2ae3eff)) and now supports aliases (0cb8184)
- Thereby the id caching mechanism (introduced in #15977) became properly generic
- The `generic_factory` doesn't expose the map object (a9a5db3) and doesn't construct a vector for iterating purposes. It returns the internal vector instead (be7deb4). Should be faster.
- The terrain objects (`ter_t`) are now managed using the factory (bea6b17). This means they can be properly modded.
- Hard-coded terrain conversion table was dropped and replaced with aliases in JSON (c998cbd).

~~WIP because `ter_t::load()` doesn't use `generic_factory` loading features yet. Also, I'm going to refactor `ter_t` checking to make it use it's own function `ter_t::check()`.~~ It's done.

### CODE REVIEW

File: src/generic_factory.h [hunk 1]
```diff
@@ -223,25 +296,63 @@ class generic_factory
...
          * casting the const away).
          */
-        const T &obj( const string_id<T> &sid ) const {
-            const auto iter = data.find( sid );
-            if( iter == data.end() ) {
-                debugmsg( "invalid %s id \"%s\"", type_name.c_str(), sid.c_str() );
-                static const T dummy{};
-                return dummy;
+        const T &obj( const string_id<T> &id ) const {
+            static int_id<T> i_id;
```
Comment (BevapDin) [inline src/generic_factory.h]: Why is this static?

File: src/generic_factory.h [hunk 2]
```diff
@@ -223,25 +296,63 @@ class generic_factory
...
+                return dummy_obj();
             }
-            return iter->second;
+            return list[i_id];
+        }
+        /**
+         * Checks whether the factory contains an object with the given id.
+         * This function can be used to implement @ref int_id::is_valid().
+         */
+        inline bool is_valid( const int_id<T> &id ) const {
```
Comment (BevapDin) [inline src/generic_factory.h]: The `inline` here is redundant. A function defined within a class definition is an automatically inline.

File: src/generic_factory.h [hunk 3]
```diff
@@ -223,25 +296,63 @@ class generic_factory
...
+            return static_cast<size_t>( id ) < list.size();
         }
         /**
          * Checks whether the factory contains an object with the given id.
          * This function can be used to implement @ref string_id::is_valid().
          */
-        bool is_valid( const string_id<T> &sid ) const {
-            return data.count( sid ) > 0;
+        bool is_valid( const string_id<T> &id ) const {
+            static int_id<T> dummy;
```
Comment (BevapDin) [inline src/generic_factory.h]: Static again, why?

File: src/generic_factory.h [hunk 4]
```diff
@@ -223,25 +296,63 @@ class generic_factory
...
-            const auto iter = data.find( sid );
-            if( iter == data.end() ) {
-                debugmsg( "invalid %s id \"%s\"", type_name.c_str(), sid.c_str() );
-                static const T dummy{};
-                return dummy;
+        const T &obj( const string_id<T> &id ) const {
+            static int_id<T> i_id;
+            if( !find_id( id, i_id ) ) {
+                debugmsg( "invalid %s id \"%s\"", type_name.c_str(), id.c_str() );
+                return dummy_obj();
```
Comment (BevapDin) [inline src/generic_factory.h]: This needs to be a static variable. Currently this returns a reference to a temporary object.

File: src/generic_factory.h (hunk 1 above)
Comment (codemime) [inline src/generic_factory.h]: To prevent constructing the object on each `obj()` invocation. It gets reassigned anyway.

File: src/generic_factory.h [hunk 5]
```diff
@@ -223,25 +296,63 @@ class generic_factory
...
          * Returns the object with the given id.
          * The input id should be valid, otherwise a debug message is issued.
+         * This function can be used to implement @ref int_id::obj().
+         * Note: If the id was valid, the returned object can be modified (after
+         * casting the const away).
+         */
+        const T &obj( const int_id<T> &id ) const {
+            if( !is_valid( id ) ) {
+                debugmsg( "invalid %s id \"%d\"", type_name.c_str(), id );
+                return dummy_obj();
```
Comment (BevapDin) [inline src/generic_factory.h]: Same here, needs to be a static variable and should return a reference to that variable.

File: src/generic_factory.h (hunk 2 above)
Comment (codemime) [inline src/generic_factory.h]: Got it.

File: src/generic_factory.h [hunk 6]
```diff
@@ -103,23 +104,83 @@ const my_class &string_id<my_class>::obj() const
...
+            if( iter == map.end() ) {
+                return false;
+            }
+            result = iter->second;
+            id.set_cid( result );
+            return true;
+        }
+
+        const T &dummy_obj() const {
+            static const T dummy;
```
Comment (BevapDin) [inline src/generic_factory.h]: Ok, I see, ignore my comments below.

But why not make this a normal member variable and return that directly?

File: src/generic_factory.h [hunk 7]
```diff
@@ -103,23 +104,83 @@ const my_class &string_id<my_class>::obj() const
...
+            if( !alias_member_name.empty() && jo.has_member( alias_member_name ) ) {
+                const int_id<T> i_id = map[id];
+                std::vector<string_id<T>> aliases;
+
+                mandatory( jo, obj.was_loaded, alias_member_name, aliases, string_id_reader<T> {} );
+
+                for( const auto &alias : aliases ) {
+                    const auto iter = map.find( alias );
+
+                    if( iter != map.end() ) {
```
Comment (BevapDin) [inline src/generic_factory.h]: You can remove the iterator and just call `map::count` to find out if the key is contained.

File: src/generic_factory.h (hunk 1 above)
Comment (BevapDin) [inline src/generic_factory.h]: But it's confusing and error prone. Constructing (and initializing) a single `int` (which is all that `int_id` contains) shouldn't take any time.

I actually suspect this will prevent the compiler from certain optimizations: a non-static variable (especially since this is effectively only an `int) can be put into a register. The static variable must be in main memory and may need to be loaded again and again.

Also: the compiler may see that id is immediately (and unconditionally) overwritten in `find` and may skip the initialization.

In short: looks like premature optimization.

File: src/generic_factory.h (hunk 6 above)
Comment (codemime) [inline src/generic_factory.h]: I tried to do so and (surprisingly) got a segfault right away. Brief investigation revealed, that assignments like `id( NULL_ID )` are sensitive to whether the specific `NULL_ID` was defined above or below the assignment. Specific `NULL_ID`s don't yet exist when in headers (for now), but compiler just doesn't care. It compiles code which fails to run. I changed the code a bit and got rid of the segfault, but it looks like escaping from the issue rather than fixing it. Ideally, broken code should not compile.

**Comment by codemime** [comment]
All set.

**Comment by Coolthulhu** [comment]
"No Fungal Mod" overrides a terrain type. Before, it would silently do it, now it complains about duplicates.

**Comment by codemime** [comment]
It should. Terrain objects probably don't have `"edit-mode" : "override"` fields.

**Comment by Coolthulhu** [comment]
No fungal mod is bundled as official mod, it needs to be handled correctly.

**Comment by codemime** [comment]
I found two terrain objects there. Fix them here or in a different PR?

**Comment by Coolthulhu** [comment]
Here

**Comment by codemime** [comment]
OK. Give me a couple of minutes.
